applied isort to seldon_core and tests

zyxue opened the pull request:
<!--
Thanks for sending a pull request! 
If this is your first time, please read our contributor guidelines:
https://docs.seldon.io/projects/seldon-core/en/latest/developer/contributing.html
-->

**What this PR does / why we need it**:

Sorted python imports in a principled way with [isort](https://pycqa.github.io/isort/), and sorted imports improve code readability.

**Which issue(s) this PR fixes**:
<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
NA

**Special notes for your reviewer**:

what's done includes:

* added `isort` to `requirements-dev.txt`
* configed `isort` in `pyproject.toml` for compatibility with black (see https://pycqa.github.io/isort/docs/configuration/black_compatibility/)
* and the run `isort tests seldon_core`

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
Otherwise, enter your extended release note in the block below.
For more information on release notes see: 
https://docs.seldon.io/projects/seldon-core/en/latest/developer/contributing.html#release-notes
-->
No, this PR shouldn't change any code logic

seldondev commented:
Hi @zyxue. Thanks for your PR.

I'm waiting for a [SeldonIO](https://github.com/orgs/SeldonIO/people) member to verify that this patch is reasonable to test. If it is, they should reply with `/ok-to-test` on its own line. Until that is done, I will not automatically test new commits in this PR, but the usual testing commands by org members will still work. Regular contributors should [join the org](https://github.com/orgs/SeldonIO/people) to skip this step.

Once the patch is verified, the new status will be reflected by the `ok-to-test` label.

I understand the commands that are listed [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository.
</details>

seldondev commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **rafalskolasinski**
You can assign the PR to them by writing `/assign @rafalskolasinski` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["rafalskolasinski"]} -->

RafalSkolasinski commented on python/seldon_core/flask_utils.py:
```diff
@@ -1,10 +1,12 @@
-from flask import request, current_app, jsonify as flask_jsonify
 from typing import Dict, Union
 
+from flask import current_app
+from flask import jsonify as flask_jsonify
+from flask import request
```
Does this need to be split into separate lines? 
If yes, I'd prefer to use `( ... )` like we have in other places.

zyxue replied:
I see. it's due to the aliasing of `jsonify as flask_jsonify`. Otherwise isort would put them into one line like
```
from flask import current_app, jsonify, request
```

RafalSkolasinski replied:
Oh, I didn't realise we had aliasing in one line :D
I guess the best is to have then
```
from flask import current_app, request
from flask import jsonify as flask_jsonify
```
?

zyxue replied:
it's because `request` comes after `jsonify` alphabetically. if it's like

```
from flask import ctx, current_app
from flask import jsonify as flask_jsonify
```
isort would keep `ctx` and `current_app` in one line.

RafalSkolasinski replied:
What if we use `# isort:skip`:
```
from flask import current_app, request
from flask import jsonify as flask_jsonify   # isort:skip
```
?

zyxue replied:
I think the main benefit of isort is that it has a set of rules for ordering imports and consistent. Perhaps we could customize then, but then it feels ad-hoc just like the current heuristic way unless we have a good reason like to be compatible with black.

I see it this way, like black, it is opinioned but consistent.

zyxue replied:
tried, I didn't know this, but isort would put a line in between
```
from flask import current_app, request

from flask import jsonify as flask_jsonify   # isort:skip
```

RafalSkolasinski replied:
Ok, cool, let's leave it like that

RafalSkolasinski replied:
And by that I meant with ` # isort:skip` ; - )

RafalSkolasinski commented:
/test integration

seldondev commented:
Thu Feb  4 17:34:05 UTC 2021
The logs for [integration] [1] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2927/1.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2927 --build=1

RafalSkolasinski commented:
BTW. I think we should also update the `fmt` or `lint` target in the `python/Makefile`. Not sure which one is better, @adriangonz? 
Also we could include installation of isort in the `setup_linter` target in the same Makefile.

adriangonz commented:
That would make a lot of sense @RafalSkolasinski. Regarding lint vs fmt, my understanding is that lint only says whether is OK or not (but doesn't modify anything), whereas fmt does modify files.

For the case of isort, we could add it on lint, to flag if the imports haven't been sorted (useful for CI). And we could add it on fmt to actually sort the imports on the file.

Does that makes sense? What are your thoughts?

RafalSkolasinski commented:
Right, that's the proper distinction! 

Sounds good, just need to keep in mind then that adding it on `lint` would need to be in form of a check that if changes are made it should fail.

adriangonz commented:
@RafalSkolasinski I think isort already has a --check-only option just for that. These tools tend to include similar options to use as linters.

RafalSkolasinski commented:
Perfecto!

@zyxue, once applying the suggestions above and passing the integration tests we will be good to merge

zyxue commented:
thanks for the clarification. when is

zyxue commented:
my bad, I misunderstood. please ignore my last comment

zyxue commented:
Why did black skip `testing/scripts`?

RafalSkolasinski commented:
it should only be skipping `testing/scripts/proto` there

RafalSkolasinski commented on python/Makefile:
```diff
@@ -95,17 +95,23 @@ push_conda:
 	black \
 		./ ../testing ../operator/helm ../operator/hack \
-		--exclude "(testing/scripts/proto|seldon_core/proto/|.eggs|.tox)"
+		--exclude "(testing|scripts|proto|seldon_core/proto/|.eggs|.tox)"
```
Should be `testing/scripts`  not `testing|scripts`

zyxue replied:
sorry for confusion, fixed.

RafalSkolasinski replied:
Hmm... actually, maybe just `scripts` is enough because this applies when being inside `../testing` already.

zyxue commented:
@RafalSkolasinski , rerun `make lint` changes these files:
```
	modified:   licenses/license.txt
	modified:   licenses/license_info.csv
	modified:   licenses/license_info.no_versions.csv
```
is that expected, should I commit these changes?

RafalSkolasinski commented:
Yeah, it probably adds `isort` licence as this is new deps.

zyxue commented:
I don't see isort in the linceses change though

zyxue commented:
@RafalSkolasinski , all tests pass now, please let me know if there is more to update.

RafalSkolasinski commented:
/test integration
/test notebooks

seldondev commented:
Thu Feb  4 22:51:35 UTC 2021
The logs for [integration] [3] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2927/3.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2927 --build=3

seldondev commented:
Thu Feb  4 22:51:35 UTC 2021
The logs for [notebooks] [2] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2927/2.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2927 --build=2

RafalSkolasinski commented on python/Makefile:
```diff
@@ -95,17 +95,23 @@ push_conda:
 	black \
 		./ ../testing ../operator/helm ../operator/hack \
-		--exclude "(testing/scripts/proto|seldon_core/proto/|.eggs|.tox)"
+		--exclude "(proto|seldon_core/proto/|.eggs|.tox)"
```
Note sure about this one, I think there was a reason why why excluded `testing/scripts`.
@axsaucedo @adriangonz?

adriangonz replied:
@RafalSkolasinski we exclude the `proto` folder within `testing/scripts`. It would be good to double check whether that expression also captures that or not. Otherwise `black` is gonna go crazy formatting the generating code for the protobufs.

RafalSkolasinski replied:
Yeah, make total sense for `protos`. Was just wondering if there is something else in `testing/scripts` we would like to leave unmoddified.

RafalSkolasinski replied:
Okay, cool, I don't see any black re-formats in the protos. Only `isort` touched some imports, added separate comment for that: https://github.com/SeldonIO/seldon-core/pull/2927#discussion_r570865145

zyxue replied:
I was thinking formatting pb2 files would be fine too. I don't see any disadvantage with that as long as it's always formatted whenever they're regenerated. But I'll follow the convention and ignore them here.

zyxue replied:
https://github.com/SeldonIO/seldon-core/tree/master/testing/scripts/proto is empty now except for a `.keep` file ...

RafalSkolasinski commented on python/seldon_core/proto/tensorflow/core/framework/__init__.py:
```diff
@@ -2,4 +2,4 @@
 if _TF_PRESENT:
     # Let tensorflow shadow these imports if present
-    from tensorflow.core.framework import tensor_pb2, types_pb2, tensor_shape_pb2
+    from tensorflow.core.framework import tensor_pb2, tensor_shape_pb2, types_pb2
```
We want to exclude protos also from `isort`.

zyxue replied:
This isn't about formatting pb2 files per se, it just formats how they're formatted

zyxue replied:
Hmm. But this is inside the proto for, let me see

RafalSkolasinski replied:
> This isn't about formatting pb2 files per se, it just formats how they're formatted

Now that's a definitely a bigger change that should go into separate issue for discussion ; - )

zyxue replied:
Sorry I meant how they're **imported***.

Ok, I agree this issue is getting overly large, I'll limit the change to seldon_core only.

zyxue replied:
I think the principal is that every manually written python files (except for auto generated ones e.g pb2 files) should follow consistent formatting enforced by black and isort, is that agreeable?

RafalSkolasinski replied:
Yes, that's right. What we write manually should be under `black` and `isort`. So all python codes except those under `proto` should fall into it. 

The tensorflow protos I think I coming from 3rd party library so we should not format them.

zyxue replied:
I see. I probably committed this one by mistake as `proto` directory has been indeed ignored. I've changed it back.

zyxue commented:
due to conflicts, I'm rebasing to master

zyxue commented:
Curious, from https://github.com/SeldonIO/seldon-core/pull/2769, I noticed that you didn't squash all commits during merge:
```
git --no-pager show --oneline --no-patch 35d9c8..f89aa4
f89aa4043 newline
f380dac57 added venv/ to .gitignore
e39767883 corrected typo
8fd5e5a29 reblacked code in seldon_core
57e6ddd32 updated black to black 20.8b1 in setup_linter
```
Is there a reason for it? I thought squashing all these commits into just one would make the master history cleaner.

RafalSkolasinski commented:
> Is there a reason for it? I thought squashing all these commits into just one would make the master history cleaner.

I don't think we are very strict on that. Also merging happens automatically by the `seldondev` bot.

zyxue commented:
> > Is there a reason for it? I thought squashing all these commits into just one would make the master history cleaner.
> 
> I don't think we are very strict on that. Also merging happens automatically by the `seldondev` bot.

ah, I see. Thanks.

I'm done with my changes now.

RafalSkolasinski commented:
/test notebooks
/test integration

seldondev commented:
Fri Feb  5 16:31:49 UTC 2021
The logs for [notebooks] [4] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2927/4.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2927 --build=4

seldondev commented:
Fri Feb  5 16:31:57 UTC 2021
The logs for [integration] [5] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2927/5.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2927 --build=5

RafalSkolasinski commented:
/approve

seldondev commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *[RafalSkolasinski](https://github.com/SeldonIO/seldon-core/pull/2927#issuecomment-774143503 "Approved")*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com#issuecomment-774143503/SeldonIO/seldon-core/blob/master/OWNERS)~~ [RafalSkolasinski]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

zyxue commented:
@RafalSkolasinski , seems blocked by `Pending — Not mergeable. Should not have needs-ok-to-test label.`
